src/pytorch_metric_learning/losses/smooth_ap.py

```diff
@@ -0,0 +1,96 @@
+import torch
+import torch.nn.functional as F
+
+from ..distances import CosineSimilarity
+from ..utils import common_functions as c_f
+from ..utils import loss_and_miner_utils as lmu
+from .base_metric_loss_function import BaseMetricLossFunction
+
+
+class SmoothAPLoss(BaseMetricLossFunction):
+    """
+    Implementation of the SmoothAP loss: https://arxiv.org/abs/2007.12163
+    """
+
+    def __init__(self, temperature=0.01, **kwargs):
+        super().__init__(**kwargs)
+        c_f.assert_distance_type(self, CosineSimilarity)
+        self.temperature = temperature
+
+    def get_default_distance(self):
+        return CosineSimilarity()
+
+    # Implementation is based on the original repository:
+    # https://github.com/Andrew-Brown1/Smooth_AP/blob/master/src/Smooth_AP_loss.py#L87
+    def compute_loss(self, embeddings, labels, iices_tuple, ref_emb, ref_labels):
```

Comment:
typo, but also is there any possibility of supporting mining, or reference embeddings (ref_emb)?

Reply:
I'm pretty sure I cannot add reference embeddings as the loss uses in-batch similarities for calculation. I've supported mining similar to the FastAP loss.